client/controller/shared.js

```diff
@@ -43,17 +101,56 @@ export function setSection( section ) {
 	};
 }
 
+export function loadSectionCSS( context, next ) {
+	const section = getSection( context.store.getState() );
+
+	if ( section.cssUrls ) {
+		const cssUrl = context.isRTL ? section.cssUrls.rtl : section.cssUrls.ltr;
+		loadCSS( cssUrl, 'section-css' );
+	}
+
+	next();
+}
+
+function fixRtlStyles() {
+	const htmlElement = document.querySelectorAll( 'html' )[ 0 ];
+
+	// this will be already set to rtl if SSR properly rendered
+	if ( htmlElement.dir === 'rtl' ) {
+		return;
+	}
+
+	const linkElements = Array.prototype.slice.apply( document.querySelectorAll( 'link[rel=stylesheet]' ) );
+	// can be style.css and style-debug.css
+	const mainStyleLink = find( linkElements, el => el.href.match( /style[^.]+\.css/ ) );
+	const mainStyleLinkHash = mainStyleLink.href.match( /\?v=([A-F0-9]+)$/i )[ 1 ];
+	const rtlStyleHref = mainStyleLink.href.split( '/' )
+		.slice( 0, -1 )
+		.concat( [ 'style-rtl.css?v=' + mainStyleLinkHash ] )
+		.join( '/' );
+
+	replaceStlyesheet( mainStyleLink, rtlStyleHref );
+
+	htmlElement.dir = 'rtl';
+}
+
 export function setUpLocale( context, next ) {
-	const { lang } = context.params;
-	const language = getLanguage( lang );
+	const currentUser = getCurrentUser( context.store.getState() );
+
+	if ( context.params.lang ) {
+		context.lang = context.params.lang;
+		context.isRTL = Boolean( getLanguage( context.lang ).rtl );
+	} else if ( currentUser ) {
+		context.lang = currentUser.localeSlug;
+		context.isRTL = isUserRtl( context.store.getState() );
 
-	if ( language ) {
-		context.lang = lang;
-		context.isRTL = Boolean( language.rtl );
+		if ( context.isRTL && ! config( 'rtl' ) ) {
+			fixRtlStyles();
```

Comment:
This will be fixed by the patch I mentioned [here](https://github.com/Automattic/wp-calypso/pull/17988#discussion_r140912465). I don't think we should be adding more "hacks" here.